### PR TITLE
Allows injection of transactions into devnet Obscuro network

### PR DESCRIPTION
**Pull request by joeldudley**
### Why is this change needed?

As part of https://github.com/obscuronet/obscuro-internal/issues/448, we need the ability to injects txs into an Obscuro devnet.

### What changes were made as part of this PR:

Functional.

* Updates `run.sh` script that runs devnet to give nodes P2P addresses (required for injecting and broadcasting txs)
* Some clean-up in transaction_injector.go:
  * Timeout if hangs for too long trying to get nonce to inject transaction
  * Handles edge cases when trying to select random client/wallet when there is only one

### What are the key areas to look at

Key changes are in the `run.sh` script.

### CODE REVIEW

File: integration/simulation/transaction_injector.go [hunk 1]
```diff
@@ -296,18 +298,30 @@ func (m *TransactionInjector) createInvalidSignage(tx types.TxData, w wallet.Wal
 }
 
 func (m *TransactionInjector) rndObsWallet() wallet.Wallet {
+	if len(m.obsWallets) == 1 {
```
Comment (joeldudley) [inline integration/simulation/transaction_injector.go:301]: `rand.Intn` spews if the len is zero, here and below.

File: integration/simulation/transaction_injector.go [hunk 2]
```diff
@@ -347,12 +361,19 @@ func readNonce(cl *obscuroclient.Client, a common.Address) uint64 {
...
 	for {
 		result := readNonce(cl, w.Address())
 		if result == w.GetNonce() {
 			return w.GetNonceAndIncrement()
 		}
 		time.Sleep(time.Millisecond)
+		counter++
+
+		if counter > timeoutMillis {
+			panic("Transaction injector failed to retrieve nonce after ten seconds...")
```
Comment (joeldudley) [inline integration/simulation/transaction_injector.go:376]: This used to spin forever if for some reason the nonce wasn't increasing (e.g. the enclave was down).

File: integration/simulation/transaction_injector.go (hunk 1 above)
Comment (BedrockSquirrel) [reply integration/simulation/transaction_injector.go:301]: I think the `rand.Intn()` usage isn't quite right here, docs says it gives an int in the half-open interval `[0, n)` so it's safe to use it like `return m.obsWallets[rand.Intn(len(m.obsWallets))]` and with the current way we'd never get the last element in the slice I think. That would let you get rid of the length 1 handling.

File: integration/simulation/transaction_injector.go (hunk 2 above)
Comment (BedrockSquirrel) [reply integration/simulation/transaction_injector.go:376]: Hitting it a thousand times a second feels a bit intense for 10 seconds. If the server is struggling under load then this is the sort of thing that causes a death spiral.

If it normally takes a couple of millis you could back off, like `time.sleep(2^retries * time.Millisecond)` and then `if (retries > 10) { panic(...) }`

File: tools/azuredeployer/networkdeployer/run.sh [hunk 3]
```diff
@@ -7,17 +7,20 @@ rm -f ./run_logs.txt
...
-sudo docker run -e OE_SIMULATION=0 --privileged -v /dev/sgx:/dev/sgx -p 127.0.0.1:11001:11000/tcp obscuro_enclave --hostID 2 --address :11000 > ./run_logs.txt 2>&1 &
-go-obscuro/go/obscuronode/host/main/host --id=1 --isGenesis=true --enclaveRPCAddress=localhost:11000 --clientRPCAddress=0.0.0.0:13000 --l1NodePort=12100 --privateKey=$PRIV_KEY_ONE > ./run_logs.txt 2>&1 &
-go-obscuro/go/obscuronode/host/main/host --id=2 --isGenesis=false --enclaveRPCAddress=localhost:11001 --clientRPCAddress=localhost:13001 --l1NodePort=12101 --privateKey=$PRIV_KEY_TWO > ./run_logs.txt 2>&1 &
+sudo docker run -e OE_SIMULATION=1 --privileged -v /dev/sgx:/dev/sgx -p 127.0.0.1:11000:11000/tcp obscuro_enclave --hostID 1 --address :11000 > ./run_logs.txt 2>&1 &
+sudo docker run -e OE_SIMULATION=1 --privileged -v /dev/sgx:/dev/sgx -p 127.0.0.1:11001:11000/tcp obscuro_enclave --hostID 2 --address :11000 > ./run_logs.txt 2>&1 &
+go-obscuro/go/obscuronode/host/main/host --id=1 --isGenesis=true --p2pAddress=localhost:10000 --peerP2PAddresses=localhost:10001 --enclaveRPCAddress=localhost:11000 --clientRPCAddress=0.0.0.0:13000 --l1NodePort=12100 --privateKey=$PRIV_KEY_ONE > ./run_logs.txt 2>&1 &
```
[the remark this answers is not in the document]
Comment (BedrockSquirrel) [reply tools/azuredeployer/networkdeployer/run.sh:23]: Worth adding a todo to remind us to remove peer addresses when they discover the peers themselves or think we'll clean it up naturally?

**Review by BedrockSquirrel — COMMENTED** [review]
Looks good, just a few minor thoughts

File: tools/azuredeployer/networkdeployer/run.sh (hunk 3 above)
[the remark this answers is not in the document]
Comment (joeldudley) [reply tools/azuredeployer/networkdeployer/run.sh:23]: Think it'll be fine. This will change a lot.

**Review by BedrockSquirrel — APPROVED** [review]
LGTM once the lint/build stuff passing. We can rework the nonce stuff in future.

**Pull request merged**